pkg/agent/api/debug/v1/service.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/spiffe/spire/pkg/agent/manager"
+	debug_pb "github.com/spiffe/spire/proto/spire/api/agent/debug/v1"
+	"github.com/spiffe/spire/proto/spire/types"
+	"github.com/spiffe/spire/test/clock"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	cacheExpiry = 5 * time.Second
+)
+
+// RegisterService registers debug service on provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	debug_pb.RegisterDebugServer(s, service)
+}
+
+// Config configurations for debug service
+type Config struct {
+	Clock       clock.Clock
+	Log         logrus.FieldLogger
+	Manager     manager.Manager
+	TrustDomain spiffeid.TrustDomain
+	Uptime      time.Time
+}
+
+// New creates a new debug service
+func New(config Config) *Service {
+	return &Service{
+		clock:  config.Clock,
+		log:    config.Log,
+		m:      config.Manager,
+		td:     config.TrustDomain,
+		uptime: config.Uptime,
+
+		getInfoResp: &getInfoResp{},
+	}
+}
+
+// Service implements debug server
+type Service struct {
+	clock  clock.Clock
+	log    logrus.FieldLogger
+	m      manager.Manager
+	td     spiffeid.TrustDomain
+	uptime time.Time
+
+	getInfoResp *getInfoResp
+}
+
+type getInfoResp struct {
+	mtx  sync.Mutex
+	resp *debug_pb.GetInfoResponse
+	ts   time.Time
+}
+
+// GetInfo gets SPIRE Agent debug information
+func (s *Service) GetInfo(ctx context.Context, req *debug_pb.GetInfoRequest) (*debug_pb.GetInfoResponse, error) {
+	s.getInfoResp.mtx.Lock()
+	defer s.getInfoResp.mtx.Unlock()
+
+	// Update cache when expired or does not exists
+	if s.getInfoResp.ts.IsZero() || s.clock.Now().Sub(s.getInfoResp.ts) >= cacheExpiry {
+		state := s.m.GetCurrentCredentials()
+		// Get current agent's credential SVID
+		svid := state.SVID
+		certChain, err := s.getCertificateChain(svid)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create SVID chain for response
+		var svidChain []*debug_pb.GetInfoResponse_Cert
+		for _, cert := range certChain {
+			svidChain = append(svidChain, &debug_pb.GetInfoResponse_Cert{
+				Id:        spiffeIDFromCert(cert),
+				ExpiresAt: cert.NotAfter.Unix(),
+				Subject:   cert.Subject.String(),
+			})
+		}
+
+		// Reset clock and set current response
+		s.getInfoResp.ts = s.clock.Now()
+		s.getInfoResp.resp = &debug_pb.GetInfoResponse{
+			SvidChain:       svidChain,
+			Uptime:          s.uptime.Unix(),
+			SvidsCount:      int32(s.m.CountSVIDs()),
+			LastSuccessSync: int32(s.m.GetLastSync().UTC().Unix()),
+		}
+	}
+
+	return s.getInfoResp.resp, nil
+}
+
+// spiffeIDFromCert gets types SPIFFE ID from certificate, it can be nil
+func spiffeIDFromCert(cert *x509.Certificate) *types.SPIFFEID {
+	id, err := x509svid.IDFromCert(cert)
+	if err != nil {
+		return nil
+	}
+
+	return &types.SPIFFEID{
+		TrustDomain: id.TrustDomain().String(),
+		Path:        id.Path(),
+	}
+}
+
+func (s *Service) getCertificateChain(svid []*x509.Certificate) ([]*x509.Certificate, error) {
```

Comment:
This is a little "sharp" to me. I think proper verification should involve the actual cached bundle and verifying using the whole partial chain present in the SVID.